rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,345 @@
+       "rule": "rule-engine",
+       "default": "default-bucket-number"
+    },
+    "<instant2>": {}
+}
+```
+Expressions has newer instant has higher priority
+
+For a specific expression like instant1 ==> Expression1 has a higher priority than expression2.
+
+Here, the config is designed to support multiple versions. During the DDL phase, hoodie will create a `00000000000000000.hashing_config` for the first time. 
+Subsequently, when modifying the expression through the provided `call command`(details refer to Tool ), 
+a new version of the config will be generated. The new-version config contains all historical change records. During 
+loading stage, the latest completed instant related config will be loaded based on the active timeline.
+
+```text
+.hoodie/.simple_hashing/config/00000000000000000.hashing_config
+.hoodie/.simple_hashing/config/20250303095546020.hashing_config
+```
+
+Using Archive Service to clean up expired config versions, keeping up to 5 historical versions
+
+### Hashing Metadata
+Here, the Hashing Metadata is designed to support multiple versions, new versions contains all the info in old versions.
+
+The hashing metadata will be persisted as files named as `<instant>.hashing_meta` for current table, it stores in
+`.hoodie/.simple_hashing/meta` directory and contains the following information
+
+```avro schema
+{
+   "namespace":"org.apache.hudi.avro.model",
+   "type":"record",
+   "doc": "hashing meta for current table using partition level simple bucket index",
+   "name":"SimpleHashMeta",
+   "fields":[
+     {
+       "name":"version",
+       "type":["int", "null"],
+       "default": 1
+     },
+     {
+       "name": "partitionMeta",
+       "type": [
+         "null", {
+           "type":"map",
+           "values": {
+             "type": "record",
+             "name": "PartitionHashInfo",
+             "fields": [
+               {
+                 "name": "bucketNumber",
+                 "type": ["null", "int"],
+                 "default": null
+               },
+               {
+                 "name": "instant",
+                 "type":["null","string"],
+                 "default": null
+               }
+             ]
+           }}],
+       "default" : null
+     }
+   ]
+}
+```
+We will write <instant>.hashing_meta during commit action (at pre-commit stage):
+1. Get latest complete T1.hashing_meta
+2. Merging T1.hashing_meta with new written partitions as T2.hashing_meta also do partition conflict detection
+   1. For example, current write write partition1 with 20 buckets but there is already defined partition1 with 10 bucket numbers, which means there is 
+another writer that uses the latest expression to pre-determine the bucket number of the current partition. In order to ensure data consistency, it is necessary to abort the current writer.
+3. If conflict detection is passed AND **new partitions are added** then write T2.hashing_meta into .hoodie folder
```

Comment:
Here is a question, if the number of partitions is large, such as hundreds of thousands, will there be performance problems in conflict detection?

Reply:
A test class that constructs 10 million pieces of data, verifies its read and write speeds, and also tests the speed of 10000 random lookups.

Time taken to write 10,000,000 records: 3,892 milliseconds.
Actual number of records read: 10,000,000.
Time taken to read 10,000,000 records: 2,453 milliseconds.
Time taken to perform 10,000 random lookups: 1 millisecond.